### ISSUE TITLE
[tflite] Output difference for simple MobileNetV2 model

### ISSUE THREAD

**Issue by hgaiser** [tensorflow/tensorflow#34642]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): **Arch Linux**
- TensorFlow installed from (source or binary): **binary**
- TensorFlow version (or github SHA if from source): **Tested on tf_nightly_gpu-2.1.0.dev20191126 from pip and 2.0.0 from the Arch repositories.**

**Command used to run the converter or code if you’re using the Python API**
Original implementation from [here](https://www.tensorflow.org/lite/convert/python_api#end-to-end_mobilenet_conversion_).

```python
import numpy as np
import tensorflow as tf

# Load the MobileNet tf.keras model (and set include_top=False).
model = tf.keras.applications.MobileNetV2(
    weights="imagenet", input_shape=(224, 224, 3), include_top=False)

# Convert the model.
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_content=tflite_model)
interpreter.allocate_tensors()

# Get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Test the TensorFlow Lite model on random input data.
input_shape = input_details[0]['shape']
input_data = np.array(np.random.random_sample(input_shape), dtype=np.float32)
interpreter.set_tensor(input_details[0]['index'], input_data)

interpreter.invoke()

# The function `get_tensor()` returns a copy of the tensor data.
# Use `tensor()` in order to get a pointer to the tensor.
tflite_results = interpreter.get_tensor(output_details[0]['index'])

# Test the TensorFlow model on random input data.
tf_results = model(tf.constant(input_data))

# Compare the result.
for tf_result, tflite_result in zip(tf_results, tflite_results):
  np.testing.assert_almost_equal(tf_result, tflite_result, decimal=5)
```

**The output from the converter invocation**

```
AssertionError: 
Arrays are not almost equal to 5 decimals

Mismatch: 2.78%
Max absolute difference: 5.4180622e-05
Max relative difference: 0.01288559
 x: array([[[0.     , 0.     , 0.     , ..., 0.     , 0.     , 0.     ],
        [0.     , 0.     , 0.     , ..., 0.     , 0.82339, 0.74743],
        [0.     , 0.88872, 0.     , ..., 0.     , 2.61463, 2.41743],...
 y: array([[[0.     , 0.     , 0.     , ..., 0.     , 0.     , 0.     ],
        [0.     , 0.     , 0.     , ..., 0.     , 0.8234 , 0.74745],
        [0.     , 0.88871, 0.     , ..., 0.     , 2.61464, 2.41744],...

```

**Failure details**
As per the [converter documentation](https://www.tensorflow.org/lite/convert/python_api), I am trying to verify that the output of my converted model and the output of my original model are identical. This is the case if I run [this](https://www.tensorflow.org/lite/convert/python_api#end-to-end_mobilenet_conversion_) code from the documentation page, however if I add `include_top=False` to the `MobileNetV2` constructor, then I get different output. I would expect the exact same output, regardless of what part of the model I select.

As per the output log, there is a max relative difference of 0.01288559, which in my opinion is pretty significant.

Note that there is no quantization being performed. It is simply creating a random MobileNetV2 model, converting it and passing through some random data. I find it weird that the output is different, but that the outputs are identical if `include_top=True`.

**Any other info / logs**
I have tested on other backbones as well, `MobileNet` and `ResNet50`; they have the same issue that the output is different when adding `include_top=False`.

**Comment by hgaiser** [comment]
Small update: I just tested it on tensorflow 1.15.0 and the same issue occurs there.

**Comment by koodzi** [comment]
Hi,
We have the same problem, and also with tflite GPU on Android.
1. Difference between "checkpoint model" and tflite inferences on a workstation 
2. Complete difference between tflite CPU and GPU inferences on Android phone (Google Pixel2, Android 9)

We managed to run our model using OpenCL and we use inverse convolution.

TF version: nightly (OpenCL), 1.15 (OpenGL), 1.13 just CPU 
TFLite compiled from source
OS Platform and Distribution: Ubuntu 18.04

UseNNAPI false
SetAllowFp16PrecisionForFp32: false

**Comment by multiverse-tf** [comment]
Many thanks for raising the issue!

As for the inference difference between tf vs tflite, it might be caused by the model conversion. @miaout17, do you have any insights into this? Thx!

As for the "complete difference between tflite CPU and GPU inferences on Android phone", @impjdi, could you shed some light on this? Thx!

**Comment by impjdi** [comment]
Google Pixel 2 doesn't support OpenCL, so I assume it's OpenGL.

@multiverse-tf 

Don't we have regression test for MobileNet v2 CPU vs GPU on Pixel 2?

**Comment by koodzi** [comment]
Below I am sharing output (audio spectrogram) of the app for the same input and parameters.
The model should denoise the audio and on CPU (upper sample) it is doing its job. And on GPU (lower sample) it is different worse (does not clear upper frequencies).

![denoised](https://user-images.githubusercontent.com/18346395/70513871-c07db500-1b32-11ea-8f07-14275c4f4cd6.png)

TFLite version 1.15 compiled from r1.15 branch 
Devices:
* Pixel  2, GPU OpenGL
* S10, GPU OpenCL

For TF compiled from master, output for both CPU and GPU is the same as 1.15 GPU output ;(

**Comment by renjie-liu** [comment]
can you run on some real dataset and check if indeed we loose some accuracy?

@multiverse-tf don't we have regression test for accuracy?

Thanks,

**Comment by multiverse-tf** [comment]
We didn't find any accuracy regression w.r.t. this MobileNetV2 model against tflite head version.

@koodzi as you built the lib against tflite 1.15, could you try the head version and see whether the issue still exists? Also, might instrument your code w/ PrintInterpreterState (i.e. tensorflow/lite/optional_debug_tools.h) to help the debugging? Thx!

**Comment by hgaiser** [comment]
Should a new issue be created for the Android issue, or is this considered the same issue but in a different use case? I'm getting somewhat confused with some of these comments:

> We didn't find any accuracy regression w.r.t. this MobileNetV2 model against tflite head version.

Are you referring to the difference in accuracy between tf and tflite **on a Pixel 2**? Or in general? If you run the provided script in the first post, does it show a difference?

**Comment by multiverse-tf** [comment]
> Should a new issue be created for the Android issue, or is this considered the same issue but in a different use case? I'm getting somewhat confused with some of these comments:
Sorry for the confusion as I also missed sth. in previous posts.
> 
> > We didn't find any accuracy regression w.r.t. this MobileNetV2 model against tflite head version.
> 
> Are you referring to the difference in accuracy between tf and tflite **on a Pixel 2**? Or in general? If you run the provided script in the first post, does it show a difference?
I mean we didn't find accuracy regressions of this model on pixel 2 using either tflite cpu or gpu if using tflite lib compiled from the head.

I haven't run the provided script yet in the first post. But I guess the difference might be caused by the different numeric computations (i.e. caused by different op kernel implementation, the model graph is different etc.) between the checkpoint model and the converted tflite model.

As for "Complete difference between tflite CPU and GPU inferences on Android phone (Google Pixel2, Android 9)", I am not sure about the cause. 

Btw, you mentioned "For TF compiled from master, output for both CPU and GPU is the same as 1.15 GPU output ;(", do you mean compiling TFLite from the github head?

**Comment by hgaiser** [comment]
> I haven't run the provided script yet in the first post. But I guess the difference might be caused by the different numeric computations (i.e. caused by different op kernel implementation, the model graph is different etc.) between the checkpoint model and the converted tflite model.

Shouldn't the outputs be nearly identical? The [converter page](https://www.tensorflow.org/lite/convert/python_api#end-to-end_mobilenet_conversion_) suggests that at least the very last output will be identical with `decimal=5` precision. I would assume, in a network where many computations are performed, that all the preceding computations have to be the same. If there is a small difference in one of the earlier layers, it would propagate throughout the network, causing a large offset at the end, but this doesn't appear to be the case.

In other words, how can MobileNetV2 **with** head produce identical results in tf and tflite, while MobileNetV2 **without** head, a subset of the complete network, produces different results? This confuses me.

**Comment by impjdi** [comment]
@hgaiser @multiverse-tf 

I think the term ``head'' is causing confusion.  When multiverse-tf was talking about tflite head version, he's talking about the latest code snapshot of TFLite repository.  When hgaiser is talking about head, it's about the bottleneck layer in the network.

Now, having said that, to hgaiser's latest comment, that statement of ``a subset of the complete network produces different results'' is only logically correct.  For GPU, we do graph modifications and tensor reshapes for performance.  If there is a bug somewhere in this process, the subset can produce wrong output, probably at the last layer (right before head) or during output tensor reshaping.  Also I would advise to leave quantization out of the game and do everything in float32 (I just learned that tf.lite.TFLiteConverter by default produces quantized models... =/ )

**Comment by impjdi** [comment]
Oh and my apologies, I wasn't paying full attention.  I see that there are two people asking for different on this thread: @hgaiser about the discrepancies of models with/without bottleneck layer and @koodzi about CPU/GPU discrepancies.  @koodzi I would advise to open up a separate thread, rather than piggybacking on someone else's thread.  Also, attaching the model can help diagnosing the issue.

Anyway, in that case, @hgaiser please ignore my comment about GPU.  It looks like either (a) there is a a bug in the converter, or (b) converter APIs are not used in the way it's intended.  I'll leave that out to @miaout17

**Comment by hgaiser** [comment]
Oh I thought that TFLiteConverter uses float32 by default. Where did you see that it uses quantisation? I would suspect that it uses float32 actually, since the differences would be greater otherwise, wouldn't it?

**Comment by impjdi** [comment]
@hgaiser 

I was referring to #34873 and it was

> converter.optimizations = [tf.lite.Optimize.DEFAULT]

which caused quantization (took me a while to find out).  If you're not using that, you should be float32 domain.

**Comment by hgaiser** [comment]
Ah I see, I didn't know that, but I'm not using it so it should be float32 then.

I ran a test where I split up the model first up until the first layer, then up until the second, third, fourth, etc. and compared the output w.r.t. the tflite converted model. For each layer I added I print if the outputs are equal or not equal. Here is the script:

```python
import numpy as np
import tensorflow as tf

# Load the MobileNet tf.keras model (and set include_top=False).
model = tf.keras.applications.MobileNetV2(
    weights="imagenet", input_shape=(224, 224, 3), include_top=True)

for layer in model.layers:
    submodel = tf.keras.Model(inputs=model.inputs, outputs=layer.output)

    # Convert the model.
    converter = tf.lite.TFLiteConverter.from_keras_model(submodel)
    tflite_model = converter.convert()

    # Load TFLite model and allocate tensors.
    interpreter = tf.lite.Interpreter(model_content=tflite_model)
    interpreter.allocate_tensors()

    # Get input and output tensors.
    input_details = interpreter.get_input_details()
    output_details = interpreter.get_output_details()

    # Test the TensorFlow Lite model on random input data.
    input_shape = input_details[0]['shape']
    input_data = np.array(np.random.random_sample(input_shape), dtype=np.float32)
    interpreter.set_tensor(input_details[0]['index'], input_data)

    interpreter.invoke()

    # The function `get_tensor()` returns a copy of the tensor data.
    # Use `tensor()` in order to get a pointer to the tensor.
    tflite_results = interpreter.get_tensor(output_details[0]['index'])

    # Test the TensorFlow model on random input data.
    tf_results = submodel(tf.constant(input_data))

    # Compare the result.
    for tf_result, tflite_result in zip(tf_results, tflite_results):
      try:
          np.testing.assert_almost_equal(tf_result, tflite_result, decimal=5)
          print(f"EQUAL {layer.__class__.__name__}")
      except:
          print(f"NOT EQUAL {layer.__class__.__name__}")
```

Here is the output:
```
EQUAL InputLayer
EQUAL ZeroPadding2D
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
NOT EQUAL ZeroPadding2D
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
EQUAL ZeroPadding2D
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
NOT EQUAL ZeroPadding2D
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
EQUAL BatchNormalization
NOT EQUAL Add
EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL ZeroPadding2D
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
EQUAL BatchNormalization
EQUAL ReLU
NOT EQUAL DepthwiseConv2D
EQUAL BatchNormalization
EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Add
NOT EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
NOT EQUAL DepthwiseConv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL Conv2D
NOT EQUAL BatchNormalization
NOT EQUAL ReLU
EQUAL GlobalAveragePooling2D
EQUAL Dense
```
I find it weird that it moves from not equal to equal, I don't really have an explanation for that.

**Comment by hgaiser** [comment]
@miaout17 do you have any idea what is happening here?

**Comment by flifuehu** [comment]
That looks to me a matter of tolerance, try decreasing `decimal` to 4 or 3

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/5c6bd5f29cc33b76ace6d0e1fb54217a/untitled90.ipynb)..Thanks !

**Comment by impjdi** [comment]
You can't really compare TF output with TFLite output, because they don't use the same implementation.  For example, floating point operations are *not* commutative, which will cause discrepancies when order of computation is done differently (a + b + c != c + b + a and a = 1e10, b = -1e10, c = 1e-10).  So don't try to compare TF with TFLite, but inspect the values manually and make a more meaningful comparison.

Also, one weird thing is that BatchNorm is preserved after conversion.  You might want to specify an option that drops the batch norm layer.

**Comment by tilakrayal** [comment]
@hgaiser ,
Please take a look at this [comment](https://github.com/tensorflow/tensorflow/issues/34642#issuecomment-852346805) and try in latest stable version2.7 and let us know if the issue still persists.Thanks!

**Comment by hgaiser** [comment]
I'm not working on this project anymore so I am unlikely to test this out. I'll close it, assuming the previous comment is correct.

**Comment by Fuze5k** [comment]
I found a solution!
https://github.com/tensorflow/tensorflow/issues/44568#issuecomment-775305887
